flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java

```diff
@@ -47,51 +46,52 @@
 import org.apache.iceberg.types.TypeUtil;
 import org.apache.iceberg.util.PartitionUtil;
 
-class RowDataIterator extends DataIterator<RowData> {
+@Internal
+public class RowDataFileScanTaskReader implements FileScanTaskReader<RowData> {
 
   private final Schema tableSchema;
   private final Schema projectedSchema;
   private final String nameMapping;
   private final boolean caseSensitive;
 
-  RowDataIterator(CombinedScanTask task, FileIO io, EncryptionManager encryption, Schema tableSchema,
-                  Schema projectedSchema, String nameMapping, boolean caseSensitive) {
-    super(task, io, encryption);
+  public RowDataFileScanTaskReader(Schema tableSchema, Schema projectedSchema,
+                                   String nameMapping, boolean caseSensitive) {
     this.tableSchema = tableSchema;
     this.projectedSchema = projectedSchema;
     this.nameMapping = nameMapping;
     this.caseSensitive = caseSensitive;
   }
 
   @Override
-  protected CloseableIterator<RowData> openTaskIterator(FileScanTask task) {
+  public CloseableIterator<RowData> open(FileScanTask task, InputFilesDecryptor inputFilesDecryptor) {
```

Comment:
If we construct the `RowDataFileScanTaskReader` inside the [DataIterator](https://github.com/apache/iceberg/pull/2905/files#diff-696823dda1232c713e658befb32b1c385a48b46c79ef25fc09edb70f5cc962a9R51),  then we don't need to pass the `InputFilesDecryptor`  for every `open` FileScanTask method, we also don't need to pass the `InputFilesDecryptor`  to the newXXXIterable methods (I mean those changes could be reverted) because they could just use the class' private `InputFilesDecryptor` instance.

Reply:
This is also tied to the other comment where we need `DataIterator` to be generic.

What you described above is similar to the current status, where we have `RowDataIterator` extends from `DataIterator`. Due to inheritance, `RowDataIterator` can call protected methods/variables from base class.

If we switch to composition model, we need to pass in `InputFilesDecryptor` to the `RowDataFileScanTaskReader`